File: crates/sync/src/stream_preview.rs
//! Inactive P1 codec. Parsing is NOT authorization; see docs/ephemeral-stream-v1.md.
//! Deliberately not registered in ChatClient or the legacy wire dispatcher.

use crate::chat_frames::{self, WireFrame};

pub const CAPABILITY: &str = "ephemeral-stream-v1";
pub const DELTA: u8 = 0x20;
pub const SNAPSHOT: u8 = 0x21;
pub const RESUME: u8 = 0x22;
pub const FINISHED: u8 = 0x23;
pub const START: u8 = 0x24;
pub const STATE: u8 = 0x25;
pub const RECEIPT: u8 = 0x26;
pub const MAX_FRAME_BYTES: usize = 65_536;
pub const MAX_TEXT_BYTES: usize = 61_440;
const MAX_SAFE_INTEGER: u64 = 9_007_199_254_740_991;

fn id(value: &serde_json::Value) -> bool {
    value.as_str().is_some_and(|s| {
        !s.is_empty()
            && s.len() <= 128
            && s.bytes()
                .all(|b| b.is_ascii_alphanumeric() || b"._:-".contains(&b))
    })
}

fn number(value: &serde_json::Value) -> Option<u64> {
    // Accept JSON 1.0 / 1e0 consistently with JavaScript and Foundation.
    let n = value.as_f64()?;
    (n >= 0.0 && n <= MAX_SAFE_INTEGER as f64 && n.fract() == 0.0).then_some(n as u64)
}

pub fn decode(bytes: &[u8]) -> Option<WireFrame> {
    if bytes.len() > MAX_FRAME_BYTES {
        return None;
    }
    let frame = chat_frames::decode(bytes)?;
    if matches!(frame.kind, START | STATE) {
        let h = frame.header.as_object()?;
        let keys: &[&str] = if frame.kind == START {
            &["chatId", "runId", "segmentId"]
        } else if h.get("mode")?.as_str()? == "preview" {
            &["chatId", "mode", "runId", "segmentId", "epoch"]
        } else {
            &["chatId", "mode"]
        };
        if !frame.payload.is_empty()
            || h.len() != keys.len()
            || !keys.iter().all(|k| h.contains_key(*k))
            || !keys.iter().filter(|k| **k != "mode").all(|k| id(&h[*k]))
            || (frame.kind == STATE
                && !matches!(h["mode"].as_str()?, "legacy" | "ready" | "preview"))
        {
            return None;
        }
        return Some(frame);
    }
    let extra = match frame.kind {
        DELTA => Some("prevRevision"),
        SNAPSHOT | RESUME | RECEIPT => None,
        FINISHED => Some("batchId"),
        _ => return None,
    };
    let header = frame.header.as_object()?;
    let common = [
        "chatId",
        "runId",
        "segmentId",
        "epoch",
        "revision",
        "baseSeq",
    ];
    if header.len() != common.len() + usize::from(extra.is_some())
        || !common.iter().all(|key| header.contains_key(*key))
        || extra.is_some_and(|key| !header.contains_key(key))
        || !common[..4].iter().all(|key| id(&header[*key]))
    {
        return None;
    }
    let revision = number(&header["revision"])?;
    number(&header["baseSeq"])?;
    match frame.kind {
        DELTA => {
            if number(&header["prevRevision"])? + 1 != revision || frame.payload.is_empty() {
                return None;
            }
        }
        FINISHED if !id(&header["batchId"]) => return None,
        _ => {}
    }
    if frame.payload.len() > MAX_TEXT_BYTES
        || std::str::from_utf8(&frame.payload).is_err()
        || (matches!(frame.kind, RESUME | FINISHED | RECEIPT) && !frame.payload.is_empty())
    {
        return None;
    }
    Some(frame)
}

pub fn encode(kind: u8, header: &serde_json::Value, text: &str) -> Option<Vec<u8>> {
    if text.len() > MAX_TEXT_BYTES {
        return None;
    }
    let bytes = chat_frames::encode(kind, header, text.as_bytes());
    decode(&bytes)?;
    Some(bytes)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::{Value, json};

    #[test]
    fn shared_vectors() {
        let vectors: Vec<Value> = serde_json::from_str(include_str!(
            "../../../edge/src/fixtures/stream-preview-v1.json"
        ))
        .unwrap();
        for v in vectors {
            let bytes = if let Some(hex) = v["hex"].as_str() {
                (0..hex.len())
                    .step_by(2)
                    .map(|i| u8::from_str_radix(&hex[i..i + 2], 16).unwrap())
                    .collect()
            } else {
                chat_frames::encode(
                    v["kind"].as_u64().unwrap() as u8,
                    &v["header"],
                    v["text"].as_str().unwrap().as_bytes(),
                )
            };
            assert_eq!(decode(&bytes).is_some(), v["valid"], "{}", v["name"]);
            if let Some(frame) = decode(&bytes) {
                let encoded = encode(
                    frame.kind,
                    &frame.header,
                    std::str::from_utf8(&frame.payload).unwrap(),
                )
                .unwrap();
                assert_eq!(decode(&encoded), Some(frame));
            }
        }
    }

    #[test]
    fn byte_limits_and_no_legacy_reinterpretation() {
        let header = json!({"chatId":"c", "runId":"r", "segmentId":"s", "epoch":"e", "revision":0, "baseSeq":0});
        assert!(encode(SNAPSHOT, &header, &"x".repeat(MAX_TEXT_BYTES)).is_some());
        assert!(encode(SNAPSHOT, &header, &"界".repeat(MAX_TEXT_BYTES / 3 + 1)).is_none());
        assert!(encode(chat_frames::frame_type::PUSH, &header, "x").is_none());
        assert!(decode(&vec![0; MAX_FRAME_BYTES + 1]).is_none());
        let mut bad_utf8 = chat_frames::encode(SNAPSHOT, &header, &[0xff]);
        assert!(decode(&bad_utf8).is_none());
        bad_utf8 = chat_frames::encode(SNAPSHOT, &json!({"pad":"x".repeat(4096)}), &[]);
        assert!(decode(&bad_utf8).is_none());
    }
}

File: apps/ios/Cypher/Sync/StreamPreview.swift
// Inactive P1 codec, not authorization. See docs/ephemeral-stream-v1.md.
// Deliberately not registered in ChatRoomClient.
import Foundation
import CoreFoundation

enum StreamPreviewWire {
    static let capability = "ephemeral-stream-v1"
    static let delta: UInt8 = 0x20
    static let snapshot: UInt8 = 0x21
    static let resume: UInt8 = 0x22
    static let finished: UInt8 = 0x23
    static let start: UInt8 = 0x24
    static let state: UInt8 = 0x25
    static let receipt: UInt8 = 0x26
    static let maxFrameBytes = 65_536
    static let maxTextBytes = 61_440

    private static func identifier(_ value: Any?) -> Bool {
        guard let value = value as? String, (1...128).contains(value.utf8.count) else { return false }
        return value.utf8.allSatisfy {
            (65...90).contains($0) || (97...122).contains($0) || (48...57).contains($0)
                || [46, 95, 58, 45].contains($0)
        }
    }

    private static func integer(_ value: Any?) -> UInt64? {
        guard let n = value as? NSNumber, CFGetTypeID(n) != CFBooleanGetTypeID() else { return nil }
        let d = n.doubleValue
        guard d.isFinite, d >= 0, d <= 9_007_199_254_740_991, d.rounded(.towardZero) == d else { return nil }
        return UInt64(d)
    }

    static func decode(_ data: Data) -> ChatWireFrame? {
        guard data.count >= 5, data.count <= maxFrameBytes else { return nil }
        let prefix = [UInt8](data.prefix(5))
        let length = Int(UInt32(prefix[1]) | UInt32(prefix[2]) << 8 | UInt32(prefix[3]) << 16 | UInt32(prefix[4]) << 24)
        guard length <= chatFrameMaxHeaderBytes, 5 + length <= data.count else { return nil }
        let rawHeader = data.subdata(in: 5..<(5 + length))
        // Foundation otherwise auto-detects UTF-16/32 and accepts a BOM, unlike
        // the UTF-8-only contract. Require an ordinary JSON object's first token.
        guard String(data: rawHeader, encoding: .utf8) != nil,
              rawHeader.first(where: { ![9, 10, 13, 32].contains($0) }) == 123,
              let frame = ChatWire.decode(data),
              [delta, snapshot, resume, finished, start, state, receipt].contains(frame.kind) else { return nil }
        if [start, state].contains(frame.kind) {
            let h = frame.header
            let keys = frame.kind == start ? ["chatId", "runId", "segmentId"]
                : h["mode"] as? String == "preview" ? ["chatId", "mode", "runId", "segmentId", "epoch"] : ["chatId", "mode"]
            guard frame.payload.isEmpty, h.count == keys.count, keys.allSatisfy({ h[$0] != nil }),
                  keys.filter({ $0 != "mode" }).allSatisfy({ identifier(h[$0]) }) else { return nil }
            if frame.kind == state && !["legacy", "ready", "preview"].contains(h["mode"] as? String ?? "") { return nil }
            return frame
        }
        var keys = ["chatId", "runId", "segmentId", "epoch", "revision", "baseSeq"]
        if frame.kind == delta { keys.append("prevRevision") }
        if frame.kind == finished { keys.append("batchId") }
        let h = frame.header
        guard h.count == keys.count, keys.allSatisfy({ h[$0] != nil }),
              keys.prefix(4).allSatisfy({ identifier(h[$0]) }),
              let revision = integer(h["revision"]), integer(h["baseSeq"]) != nil,
              frame.payload.count <= maxTextBytes,
              String(data: frame.payload, encoding: .utf8) != nil else { return nil }
        if frame.kind == delta {
            guard let previous = integer(h["prevRevision"]), previous + 1 == revision,
                  !frame.payload.isEmpty else { return nil }
        }
        if frame.kind == finished && !identifier(h["batchId"]) { return nil }
        if [resume, finished, receipt].contains(frame.kind) && !frame.payload.isEmpty { return nil }
        return frame
    }

    static func encode(_ kind: UInt8, header: [String: Any], text: String = "") -> Data? {
        guard text.utf8.count <= maxTextBytes, JSONSerialization.isValidJSONObject(header) else { return nil }
        let bytes = ChatWire.encode(kind, header: header, payload: Data(text.utf8))
        return decode(bytes) == nil ? nil : bytes
    }
}
